README.md
[![License][license-image]][license]
[![Coverage Status][coveralls-image]][coveralls]
[![Test][test-image]][test]
[![Docker][docker-image]][docker]

# Snowplow Postgres Loader

## Overview

The Snowplow Postgres Loader consumes records from an Amazon Kinesis stream and Google Pubsub topic and writes them to PostgreSQL database.

## Find out more

| Technical Docs             | Setup Guide           | Roadmap              | Contributing                |
|:--------------------------:|:---------------------:|:--------------------:|:---------------------------:|
| ![i1][techdocs-image]      | ![i2][setup-image]    | ![i3][roadmap-image] |![i4][contributing-image]    |
| [Technical Docs][techdocs] | [Setup Guide][config] | [Roadmap][roadmap]   |[Contributing][contributing] |


## Copyright and License

Snowplow Postgres Loader is copyright 2020-2021 Snowplow Analytics Ltd.

Licensed under the **[Apache License, Version 2.0][license]** (the "License");
you may not use this software except in compliance with the License.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[config]: https://docs.snowplow.io/docs/pipeline-components-and-applications/loaders-storage-targets/snowplow-postgres-loader/postgres-loader-configuration-reference/
[techdocs]: https://docs.snowplow.io/docs/pipeline-components-and-applications/loaders-storage-targets/snowplow-postgres-loader/
[roadmap]: https://github.com/snowplow/snowplow/projects/7
[contributing]: https://docs.snowplow.io/docs/contributing/

[docker]: https://hub.docker.com/r/snowplow/snowplow-postgres-loader/tags
[docker-image]: https://img.shields.io/docker/v/snowplow/snowplow-postgres-loader/latest

[test]: https://github.com/snowplow-incubator/snowplow-postgres-loader/actions?query=workflow%3ATest
[test-image]: https://github.com/snowplow-incubator/snowplow-postgres-loader/workflows/Test/badge.svg

[license]: http://www.apache.org/licenses/LICENSE-2.0
[license-image]: http://img.shields.io/badge/license-Apache--2-blue.svg?style=flat

[coveralls]: https://coveralls.io/github/snowplow-incubator/snowplow-postgres-loader?branch=master
[coveralls-image]: https://coveralls.io/repos/github/snowplow-incubator/snowplow-postgres-loader/badge.svg?branch=master

[techdocs-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/techdocs.png
[setup-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/setup.png
[roadmap-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/roadmap.png
[contributing-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/contributing.png


CHANGELOG
Version 0.3.3 (2022-09-30)
--------------------------
Bump apache commons-compress to 1.21 (#76)
Bump netty-codec to 4.1.68 (#76)
Bump protobuf-java to 3.16.1 (#76)
Bump scala version to 2.13.9 (#76)
Bump gson to 2.8.9 (#76)
Exclude kafka-clients library (#76)
Bump postgresql driver to 42.4.1 (#76)
Bump jackson-databind to 2.12.6.1 (#76)

Version 0.3.2 (2021-11-15)
--------------------------
Use direct executor for KPL callback (#73)
Github action to scan docker image with lacework (#72)
Warn about bad rows even in DummyStream sink (#64)
Add latest tag to Docker build explicitly (#61)
Disable colored logging by default (#63)
Fix flaky sink test: return columns in order (#71)
Checkpoint outstanding records when loader terminates (#69)
Fix timestamp in parsing error (#68)
Use ContextShift after Async callback in KinesisSink (#67)

Version 0.3.1 (2021-08-30)
--------------------------
Make database thread pool size configurable (#60)

Version 0.3.0 (2021-08-27)
--------------------------
Use a local directory as an event source (#42)
Pass stream name and application name to fanout config object (#44)
Improve checkpointing of kinesis records (#39)
Bump sbt to 1.5.5 (#46)
Bump postgresql to 42.2.23 (#47)
Bump logback to 1.2.5 (#48)
Bump decline to 2.1.0 (#49)
Bump cats-effect to 2.5.3 (#50)
Bump fs2 to 2.5.9 (#51)
Bump snowplow-badrows to 2.1.1 (#52)
Bump schema-ddl to 0.14.1 (#53)
Send bad rows to separate stream (#54)
Make database max pool size configurable (#56)
Prepare schema definition to work with pii pseudoanonymization (#45)
Run tests of different modules seperately in CI/CD (#58)

Version 0.2.0 (2021-07-08)
--------------------------
Improve configuration defaults and examples (#41)
Remove unused name and id fields from config hocon (#38)
Ignore missing comment on atomic events table (#36)
Add PubSub configuration to the example config file (#34)
Rename storage to output in config file (#33)
Use jsonb instead of varchar for complex enums (#31)
Add ability to disable CloudWatch metric reportings (#11)
Remove requirement to always use Kinesis Enhanced Fan Out (#14)
Bump postgresql to 42.2.0 (#15)
Bump schema-ddl to 0.13.1 (#27)
Bump log4s to 1.10.0 (#26)
Bump fs2 to 2.5.6 (#25)
Bump doobie to 0.13.4 (#24)
Bump fs2-google-pubsub to 0.17.0 (#23)
Bump fs2 aws to 3.1.1 (#22)
Bump circe to 0.14.1 (#30)
Bump cats-effect to 2.5.1 (#29)
Bump decline to 2.0.0 (#28)
Bump commons-codec to 1.15 (#21)
Use hocon for cli config instead of self-describing json (#20)
Build docker images for amd64 and arm64 architectures (#19)
Use AdoptOpenJDK 11 as docker base image (#18)
Publish to maven central instead of bintray (#17)
Attach jar files to github release (#10)
Fix type derivation of enum fields (#12)

Version 0.1.1 (2020-11-10)
--------------------------
Add support for nullable fields (#6)
Bump snowplow-scala-analytics-sdk to 2.1.0 (#8)

Version 0.1.0 (2020-10-05)
--------------------------
Initial release


project/Dependencies.scala
/*
 * Copyright (c) 2020 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */

import sbt._

object Dependencies {

  object V {
    // Java
    val postgres        = "42.4.1"
    val commons         = "1.15"
    val logback         = "1.2.5"
    val jacksonDatabind = "2.12.6.1"
    val gson            = "2.8.9"
    val protobuf        = "3.16.1"
    val nettyCodec      = "4.1.68.Final"
    val commonsCompress = "1.21"

    // Scala third-party
    val decline      = "2.1.0"
    val catsEffect   = "2.5.3"
    val circe        = "0.14.1"
    val circeConfig  = "0.8.0"
    val fs2Aws       = "3.1.1"
    val fs2PubSub    = "0.17.0"
    val doobie       = "0.13.4"
    val fs2          = "2.5.9"
    val log4s        = "1.10.0"
    val config       = "1.4.1"
    val fs2BlobCore  = "0.7.3"
    val catsRetry    = "2.1.1"

    val analyticsSdk = "2.1.0"
    val badRows      = "2.1.1"
    val schemaDdl    = "0.14.1"
    val igluClient   = "1.1.1"

    // Testing
    val specs2       = "4.9.4"
    val scalaCheck   = "1.14.3"
  }

  // Java
  val logback         = "ch.qos.logback"             % "logback-classic"  % V.logback
  val config          = "com.typesafe"               % "config"           % V.config
  val jacksonDatabind = "com.fasterxml.jackson.core" % "jackson-databind" % V.jacksonDatabind // override transitive version to address security vulnerabilities
  val gson            = "com.google.code.gson"       % "gson"             % V.gson // override transitive version to address security vulnerabilities
  val protobuf        = "com.google.protobuf"        % "protobuf-java"    % V.protobuf // override transitive version to address security vulnerabilities
  val nettyCodec      = "io.netty"                   % "netty-codec"      % V.nettyCodec
  val commonsCompress = "org.apache.commons"         % "commons-compress" % V.commonsCompress

  // Snyk warnings
  val postgres      = "org.postgresql" % "postgresql"             % V.postgres
  val commons       = "commons-codec"  % "commons-codec"          % V.commons

  // Scala third-party
  val catsEffect    = "org.typelevel"            %% "cats-effect"            % V.catsEffect
  val decline       = "com.monovore"             %% "decline"                % V.decline
  val fs2           = "co.fs2"                   %% "fs2-core"               % V.fs2
  val fs2Io         = "co.fs2"                   %% "fs2-io"                 % V.fs2
  val circe         = "io.circe"                 %% "circe-core"             % V.circe
  val circeGeneric  = "io.circe"                 %% "circe-generic"          % V.circe
  val circeExtras   = "io.circe"                 %% "circe-generic-extras"   % V.circe
  val circeParser   = "io.circe"                 %% "circe-parser"           % V.circe
  val circeConfig   = "io.circe"                 %% "circe-config"           % V.circeConfig
  val circeLiteral  = "io.circe"                 %% "circe-literal"          % V.circe
  val fs2Aws        = "io.laserdisc"             %% "fs2-aws"                % V.fs2Aws
  val fs2PubSub     = "com.permutive"            %% "fs2-google-pubsub-grpc" % V.fs2PubSub
  val doobie        = "org.tpolecat"             %% "doobie-core"            % V.doobie
  val doobiePg      = "org.tpolecat"             %% "doobie-postgres"        % V.doobie
  val doobiePgCirce = "org.tpolecat"             %% "doobie-postgres-circe"  % V.doobie
  val doobieHikari  = "org.tpolecat"             %% "doobie-hikari"          % V.doobie
  val log4s         = "org.log4s"                %% "log4s"                  % V.log4s
  val fs2BlobCore   = "com.github.fs2-blobstore" %% "core"                   % V.fs2BlobCore
  val catsRetry     = "com.github.cb372"         %% "cats-retry"             % V.catsRetry

  // Scala first-party
  val analyticsSdk = "com.snowplowanalytics" %% "snowplow-scala-analytics-sdk" % V.analyticsSdk
  val badRows      = "com.snowplowanalytics" %% "snowplow-badrows"             % V.badRows
  val schemaDdl    = "com.snowplowanalytics" %% "schema-ddl"                   % V.schemaDdl
  val igluClient   = "com.snowplowanalytics" %% "iglu-scala-client"            % V.igluClient

  // Testing
  val specs2       = "org.specs2"     %% "specs2-core"       % V.specs2     % Test
  val specs2Check  = "org.specs2"     %% "specs2-scalacheck" % V.specs2     % Test
  val scalaCheck   = "org.scalacheck" %% "scalacheck"        % V.scalaCheck % Test

  val exclusions = Seq(
    ExclusionRule(organization = "org.apache.kafka", name = "kafka-clients")
  )
}
